Paperqa impl

iQuxLE opened the pull request:
linked to #73

- basic starting impl of paperqa, we will build on top of this

justaddcoffee commented:
@iQuxLE cool! Works on my end 

Do we need an "ask" command in paperqa_cli.py?

justaddcoffee commented:
(to ask questions using a group of indexed papers)

iQuxLE commented:
In discussion with @justaddcoffee we want to have it that the index is directly where the papers are, in a subfolder `.pqa` - and this works from the cli commands with the recent changes.

However we broke the tools using this. So with the recent changes you cannot use the agent. I'm trying to fix this now.

iQuxLE commented:
_Ready for Review_

  #### Implementation Details

  - **CLI Commands**: The `paperqa_cli.py` implementation provides commands for:
    - `aurelian paperqa index`: Index papers in a directory
    - `aurelian paperqa list`: List indexed papers
    - `aurelian paperqa ask`: Ask questions about papers
    - All CLI commands set `PQA_HOME` to match the paper directory, keeping indexes and papers together

  - **Agent Integration**:
    - Implemented in the standard Aurelian agent pattern with `_agent.py`, `_tools.py`, `_config.py`, and `_gradio.py`
    - Added error handling for empty indexes and more user-friendly messages
    - Accessible via `aurelian agent --agent paperqa --ui`

  - **Tool Functions**:
    - `search_papers`: Find papers based on keywords
    - `query_papers`: Ask questions about indexed papers
    - `add_paper`: Add a paper from URL or file path
    - `list_papers`: List available papers
    - `build_index`: Rebuild search index

  ## Usage Notes

  1. **CLI vs. Gradio**:
     - The CLI is optimized for indexing operations with explicit control over index locations
     - The Gradio UI now sets `PQA_HOME` to match the paper directory for consistency

  2. **Index Management**:
     - By default, indexes are stored with papers when using CLI commands
     - The same pattern is now applied in the Gradio interface

  All functionality is working as expected, though the formatting in the Gradio output is more limited compared to the CLI output (this appears to be a limitation of how the agent tools output is rendered).

  @cmungall @justaddcoffee @caufieldjh

justaddcoffee commented:
@iQuxLE I'm seeing this error when indexing papers:

```
(aurelian) ~/PythonProject/aurelian paperqa_impl $ poetry run aurelian paperqa index -d tests/data/3_pdfs/
Found 3 PDF files in tests/data/3_pdfs/
Index will be stored in: tests/data/3_pdfs/
Indexing papers... (this may take a while)
Logfire project URL: https://logfire-us.pydantic.dev/justaddcoffee/starter-project
Error indexing papers: Error -5 while decompressing data: incomplete or truncated stream
```

Can you confirm that this happens on your end?

justaddcoffee commented:
@iQuxLE this might fix the above issue

caufieldjh commented:
_very_ minor note, and perhaps a subject for a different issue, but running `aurelian paperqa index` without an api key set produces a very messy stack trace as the error propagates through `openai`, `litellm`, `lmi`, and `paperqa`.

caufieldjh commented:
Some additional strangeness: The `aurelian paperqa index` command defaults to the current working directory, so if I'm in the root of the cloned `aurelian` repo, that's great - I can just move PDFs there. But then the command appears to index the local PDFs _and_ projects docs.
I get many errors like `Error parsing docs/agents/checklist_agent.md, skipping index for this file.`
At the conclusion of the indexing, I get the message `Success! 77 papers indexed.`
This does not appear to be true - if I run `aurelian paperqa list` then the output only includes the local PDFs.
e.g., I have just one PDF named `ECID.pdf` in the project root directory, and after indexing, the `list` command says:
```
Files in paper directory /home/harry/aurelian:
  - ECID.pdf

No indexed papers found. Run 'aurelian paperqa index' to index papers.
Logfire project URL: https://logfire-us.pydantic.dev/caufieldjh/starter-project
```

And then if I run the `aurelian paperqa ask` command, it fails:
```
Error querying papers: Index pqa_index_057cda80d964ea45e95c11e798041d7a was empty, please rebuild it.
Logfire project URL: https://logfire-us.pydantic.dev/caufieldjh/starter-project
```

caufieldjh commented:
The `index` and `list` commands work more like expected if I specify a directory with the `--directory` flag, though the list command still says `No indexed papers found. Run 'aurelian paperqa index' to index papers.` even when there is a listed, indexed paper.

caufieldjh commented:
But then when I run the `ask` command with the `--directory` flag and point it to the same directory as in the `index` and `list` commands, I'm told the index is empty, *unless* I also specify the index location in a command like
```
aurelian paperqa ask --directory pqa_source/ --index-location pqa_source/.pqa/indexes/ "What is ECID?"
```

justaddcoffee commented:
Thanks Harry - I think this is broken. IMHO the code should behave like this:

User should have to specify the paper directory `-d` for all paperqa_cli commands. This should be required and not have a default, and all commands should die loudly if `-d` is not specified. There should not be an index location `-i` arg, and the index location should automatically be paper directory + ".pqa/indexes". 

IMHO paperqa_cli commands `ask()` and `index()` should also die loudly if OPENAI_API_KEY is not set

iQuxLE commented:
@justaddcoffee 
> @iQuxLE I'm seeing this error when indexing papers:
> 
> ```
> (aurelian) ~/PythonProject/aurelian paperqa_impl $ poetry run aurelian paperqa index -d tests/data/3_pdfs/
> Found 3 PDF files in tests/data/3_pdfs/
> Index will be stored in: tests/data/3_pdfs/
> Indexing papers... (this may take a while)
> Logfire project URL: https://logfire-us.pydantic.dev/justaddcoffee/starter-project
> Error indexing papers: Error -5 while decompressing data: incomplete or truncated stream
> ```
> 
> Can you confirm that this happens on your end?

I cant confirm but I will iterate on this, currently I am getting this:
```

>(aurelian-py3.11) aurelian % poetry run aurelian paperqa index -d tests/data/3_pdfs/
Found 4 PDF files in tests/data/3_pdfs/
Index will be stored in: tests/data/3_pdfs/
Indexing papers... (this may take a while)
Logfire project URL: https://logfire-us.pydantic.dev/iquxle/aurelian
CROSSREF_MAILTO environment variable not set. Crossref API rate limits may apply.
CROSSREF_API_KEY environment variable not set. Crossref API rate limits may apply.
SEMANTIC_SCHOLAR_API_KEY environment variable not set. Semantic Scholar API rate limits may apply.
SEMANTIC_SCHOLAR_API_KEY environment variable not set. Semantic Scholar API rate limits may apply.
SEMANTIC_SCHOLAR_API_KEY environment variable not set. Semantic Scholar API rate limits may apply.
SEMANTIC_SCHOLAR_API_KEY environment variable not set. Semantic Scholar API rate limits may apply.
Success! 4 papers indexed.

```

iQuxLE commented:
Thanks for sharing issues!
